flink-runtime/src/main/java/org/apache/flink/runtime/rest/handler/AbstractRestHandler.java

```diff
@@ -0,0 +1,201 @@
+import org.apache.flink.shaded.netty4.io.netty.buffer.Unpooled;
+import org.apache.flink.shaded.netty4.io.netty.channel.ChannelFuture;
+import org.apache.flink.shaded.netty4.io.netty.channel.ChannelFutureListener;
+import org.apache.flink.shaded.netty4.io.netty.channel.ChannelHandler;
+import org.apache.flink.shaded.netty4.io.netty.channel.ChannelHandlerContext;
+import org.apache.flink.shaded.netty4.io.netty.channel.SimpleChannelInboundHandler;
+import org.apache.flink.shaded.netty4.io.netty.handler.codec.http.DefaultHttpResponse;
+import org.apache.flink.shaded.netty4.io.netty.handler.codec.http.FullHttpRequest;
+import org.apache.flink.shaded.netty4.io.netty.handler.codec.http.HttpHeaders;
+import org.apache.flink.shaded.netty4.io.netty.handler.codec.http.HttpRequest;
+import org.apache.flink.shaded.netty4.io.netty.handler.codec.http.HttpResponse;
+import org.apache.flink.shaded.netty4.io.netty.handler.codec.http.HttpResponseStatus;
+import org.apache.flink.shaded.netty4.io.netty.handler.codec.http.LastHttpContent;
+import org.apache.flink.shaded.netty4.io.netty.handler.codec.http.router.Routed;
+
+import com.fasterxml.jackson.core.JsonParseException;
+import com.fasterxml.jackson.databind.JsonMappingException;
+import com.fasterxml.jackson.databind.ObjectMapper;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import javax.annotation.Nonnull;
+
+import java.io.IOException;
+import java.io.StringWriter;
+import java.util.concurrent.CompletableFuture;
+
+import static org.apache.flink.shaded.netty4.io.netty.handler.codec.http.HttpHeaders.Names.CONNECTION;
+import static org.apache.flink.shaded.netty4.io.netty.handler.codec.http.HttpHeaders.Names.CONTENT_TYPE;
+import static org.apache.flink.shaded.netty4.io.netty.handler.codec.http.HttpVersion.HTTP_1_1;
+
+/**
+ * Super class for netty-based handlers that work with {@link RequestBody}s and {@link ResponseBody}s.
+ *
+ * <p>Subclasses must be thread-safe.
+ *
+ * @param <R> type of incoming requests
+ * @param <P> type of outgoing responses
+ */
+@ChannelHandler.Sharable
+public abstract class AbstractRestHandler<R extends RequestBody, P extends ResponseBody> extends SimpleChannelInboundHandler<Routed> {
+	protected final Logger log = LoggerFactory.getLogger(getClass());
+
+	private static final ObjectMapper mapper = RestMapperUtils.getStrictObjectMapper();
+
+	private final MessageHeaders<R, P, ?> messageHeaders;
+
+	protected AbstractRestHandler(MessageHeaders<R, P, ?> messageHeaders) {
+		this.messageHeaders = messageHeaders;
+	}
+
+	public MessageHeaders<R, P, ?> getMessageHeaders() {
+		return messageHeaders;
+	}
+
+	@Override
+	protected void channelRead0(final ChannelHandlerContext ctx, Routed routed) throws Exception {
+		log.debug("Received request.");
+		final HttpRequest httpRequest = routed.request();
+
+		try {
+			if (!(httpRequest instanceof FullHttpRequest)) {
+				log.error("Implementation error: Received a request that wasn't a FullHttpResponse.");
```

Comment:
It's not that they are not allowed, but we use a message aggregator in the pipeline that always returns a `FullHttpRequest`.